libraries/chain/backing_store/db_key_value_any_lookup.cpp

```diff
@@ -0,0 +1,128 @@
+#include <eosio/chain/apply_context.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/backing_store/db_key_value_any_lookup.hpp>
+#include <eosio/chain/backing_store/db_key_value_format.hpp>
+#include <eosio/chain/backing_store/db_context.hpp>
+#include <eosio/chain/backing_store/chain_kv_payer.hpp>
+#include <eosio/chain/combined_database.hpp>
+
+namespace eosio { namespace chain { namespace backing_store {
+
+   eosio::session::shared_bytes make_useless_value() {
+      const char null = '\0';
+      return eosio::session::make_shared_bytes(&null, 1);
+   }
+
+   const eosio::session::shared_bytes db_key_value_any_lookup::useless_value = make_useless_value();
+
+   void db_key_value_any_lookup::add_table_if_needed(const shared_bytes& key, account_name payer) {
+      auto table_key = db_key_value_format::create_full_key_prefix(key, end_of_prefix::pre_type);
+      auto session_iter = current_session.lower_bound(table_key);
+      if (!match_prefix(table_key, session_iter)) {
+         // need to add the key_type::table to the end
+         table_key = db_key_value_format::create_table_key(table_key);
+         const auto legacy_key = db_key_value_format::extract_legacy_key(table_key);
+         const auto extracted_data = db_key_value_format::get_prefix_thru_key_type(legacy_key);
+         std::string event_id;
+         apply_context& context = parent.context;
+         const auto& scope = std::get<0>(extracted_data);
+         const auto& table = std::get<1>(extracted_data);
+         auto dm_logger = context.control.get_deep_mind_logger();
+         if (dm_logger != nullptr) {
+            event_id = db_context::table_event(parent.receiver, scope, table);
+         }
+
+         context.update_db_usage(payer, table_overhead, db_context::add_table_trace(context.get_action_id(), event_id.c_str()));
+
+         payer_payload pp(payer, nullptr, 0);
+         current_session.write(table_key, pp.as_payload());
+
+         if (dm_logger != nullptr) {
+            db_context::write_insert_table(*dm_logger, context.get_action_id(), parent.receiver, scope, table, payer);
+         }
+      }
+   }
+
+   void db_key_value_any_lookup::remove_table_if_empty(const shared_bytes& key) {
+      // look for any other entries in the table
+      auto entire_table_prefix_key = db_key_value_format::create_full_key_prefix(key, end_of_prefix::at_type);
+      // since this prefix key is just scope and table, it will include all primary, secondary, and table keys
+      auto session_itr = current_session.lower_bound(entire_table_prefix_key);
+      EOS_ASSERT( session_itr != current_session.end(), db_rocksdb_invalid_operation_exception,
```

Comment:
Something that you might want to see if it makes a difference in performance is checking `session_itr.key()` instead.  The end iterator will return an empty key and if you have an iterator that returns an empty key, it must be the end.  That would bypass the needed creation of the end iterator.  But I'm assuming that keys can never be empty.

Reply:
well, for any of my code the key should at the very least have the db type prefix, so empty key does mean we are not looking at anything in the db intrinsics portion of the code.